python/ml4ir/base/model/calibration/temperature_scaling.py

```diff
@@ -0,0 +1,329 @@
+
+        return tf.reduce_mean(
+            tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels_tensor,
+                                                           logits=logits_w_temp))
+
+    @tf.function
+    def nll_with_lbfgs():
+        """Returns optimizer function. Inspired by
+         https://github.com/gpleiss/temperature_scaling/blob/master/temperature_scaling.py#L60"""
+        return tfp.optimizer.lbfgs_minimize(make_val_and_grad_fn(nll_loss_fn),
+                                            initial_position=start)
+
+    # evaluation on validation set before temperature scaling
+    logits_numpys, labels_numpys = get_logits_labels(intermediate_model, dataset.validation)
+    original_acc_op, original_nll_loss_op, _ = eval_relevance_model(scorer, logits_numpys,
+                                                                    labels_numpys)
+
+    # perform temperature scaling
+    results = run_optimizer(nll_with_lbfgs, logger)
+
+    # evaluation on validation set with temperature scaling
+    temper = tf.constant(results.position)
+    acc_op, _, _ = eval_relevance_model(scorer, logits_numpys, labels_numpys, temperature=temper)
+
+    logger.info("=" * 50)
+    logger.info(f'temperature value : {results.position}')
+    logger.info('Evaluating on the validation dataset ')
+    logger.info(f'original loss: {original_nll_loss_op}, accuracy: {original_acc_op}, \n'
+          f'temperature scaling loss: {results.objective_value}, accuracy: {acc_op}\n')
+    logger.info("="*50)
+    logger.info("Evaluating on the test dataset")
+
+    logits_numpys_test, labels_numpys_test = get_logits_labels(intermediate_model, dataset.test)
+
+    # evaluation on test set before temperature scaling
+    acc_test_original, _, original_softmaxes = eval_relevance_model(scorer, logits_numpys_test,
+                                                                    labels_numpys_test)
+
+    # evaluation on test set with temperature scaling
+    acc_test_temperature_scaling, _, temperature_scaling_softmaxes =\
+        eval_relevance_model(scorer, logits_numpys_test, labels_numpys_test, temperature=temper)
+
+    # write the full vector in the csv not ...
+    np.set_printoptions(
+        formatter={'all': lambda x: str(x.decode('utf-8')) if isinstance(x, bytes) else str(x)},
+        linewidth=sys.maxsize,
+        threshold=sys.maxsize)
+    # avoiding .tolist() that is not memory efficient
+    # note: temperature scaling does not change the accuracy as it does not change the maximum. So,
+    # the temperature scaling predicted labels must be the same as  original
+    # predicted labels: `original_predicted_label`
+    data_dic = {'original_scores': [x for x in original_softmaxes.numpy()],
+                'temperature_scaling_scores': [x for x in temperature_scaling_softmaxes.numpy()],
+                'original_predicted_label': tf.argmax(original_softmaxes, axis=-1).numpy(),
```

Comment:
Perhaps calling `scores` -> `probabilities` for future? Thanks for the changes though!